R/imputation.R

```diff
@@ -0,0 +1,326 @@
+      "MinDet", "MinProb", "min", "zero",
+      "mixed", "nbavg", "none")
+
+##' @export
+##' @rdname imputation
+##' @param k `numeric(1)` providing the imputation value used for the
+##'     first and last samples if they contain an `NA`. The default is
+##'     to use the smallest value in the data.
+impute_neighbour_average <- function(x, k = NULL) {    
+    message("Assuming values are ordered.")
+    if (is.null(k))
+        k <- min(x, na.rm = TRUE)
+    .Call('_MsCoreUtils_imp_neighbour_avg', x, k)
+}
+
+##' @export
+##' @rdname imputation
+impute_knn <- function(x, ...) {
+    requireNamespace("impute")        
+    imp_res <- impute::impute.knn(x, ...)
+    if (!is.null(imp_res$rng.state)) {
+        assign(".Random.seed", imp_res$rng.state, envir = .GlobalEnv)
+    } else {
+        rm(".Random.seed", envir = .GlobalEnv)
+    }
+    imp_res$data
+}
+
+##' @export
+##' @rdname imputation
+impute_mle <- function(x, ...) {
+    requireNamespace("norm")
+    s <- norm::prelim.norm(x)  ## preliminary manipulations
+    th <- norm::em.norm(s, ...) ## find the MLE
+    seed <- sample(.Machine$integer.max, 1)
+    norm::rngseed(seed) ## set random number generator seed
+    norm::imp.norm(s, th, x)  ## impute missing data under the MLE
+}
+
+##' @export
+##' @rdname imputation
+impute_bpca <- function(x, ...) {
+    requireNamespace("pcaMethods")
+    nSamples <- dim(x)[2]
+    .resultBPCA <- pcaMethods::pca(x,
+                                   method = "bpca",
+                                   nPcs = (nSamples - 1),
+                                   verbose = FALSE,
+                                   ...)
+    pcaMethods::completeObs(.resultBPCA)
+}
+
+##' @param randna `logical` of length equal to `nrow(object)` defining
+##'     which rows are missing at random. The other ones are
+##'     considered missing not at random. Only relevant when `methods`
+##'     is `mixed`.
+##' 
+##' @param mar Imputation method for values missing at random. See
+##'     `method` above.
+##' 
+##' @param mnar Imputation method for values missing not at
+##'     random. See `method` above.
+##'
+##' @export
+##' 
+##' @rdname imputation
+impute_mixed <- function(x, randna, mar, mnar, ...) {
+    if (missing(randna))
+        stop("Mixed imputation requires 'randna' argument. See ?impute_mixed.",
+             call. = FALSE)
+    stopifnot(is.logical(randna))
+    if (missing(mar))
+        stop("Mixed imputation requires 'mar' argument. See ?impute_mixed.",
+             call. = FALSE)
+    if (missing(mnar))
+        stop("Mixed imputation requires 'mnar' argument. See ?impute_mixed.",
+             call. = FALSE)
+    if (length(randna) != nrow(x))
+        stop("Number of proteins and length of randna must be equal.",
+             call. = FALSE)
+    x[randna, ] <- impute_matrix(x[randna, ], mar, ...)
+    x[!randna, ] <- impute_matrix(x[!randna, ], mnar, ...)
+    x
+}
+
+##' @export
+##' @rdname imputation
+impute_min <- function(x) {
```

Comment:
Would we also want to allow users to replace `NA` also with only a fraction/proportion of the min? Something like

```r
impute_min <- function(x, prop = 1) {
    val <- min(x, na.rm = TRUE) * prop
    x[is.na(x)] <- val
    x
}
```

so it would also be possible to replace `NA`s with half the minimum `impute_min(x, prop = 0.5)`.

Reply:
I am not sure about that. I would then suggest a `impute_with`, and users can set to whatever they want, for example `min(x, na.rm = TRUE) * 0.5`. I will also add that one as an example.